ska_trend/bad_periscope_gradient/periscope_update.py

```diff
@@ -0,0 +1,160 @@
+from pathlib import Path
+
+import numpy as np
+from astropy.table import Table, vstack
+from cheta import fetch
+from cxotime import CxoTime
+from kadi import events
+from ska_helpers.logging import basic_logger
+
+from ska_trend import __version__
+
+logger = basic_logger(__name__, level="INFO")
+
+
+def get_opt():
+    parser = argparse.ArgumentParser(
+        description="Periscope gradient checker {}".format(__version__)
+    )
+    parser.add_argument(
+        "--email",
+        type=str,
+        default=[],
+        action="append",
+        dest="emails",
+        help="Email address to send alerts",
+    )
+    parser.add_argument("--out-dir", type=str, default=".", help="Out directory")
+    return parser
+
+
+def check_for_bad_times(start):
+    """
+    Review periscope gradient data for discontinuities and bad data.
+
+    This uses 0.0032 K as the quantization and checks that the difference between
+    consecutive samples are not more than 5 * quantization.  Times with samples
+    that exceed this threshold are put into a table along with the telemetered
+    COBSRQID, AOACASEQ, and AOPCADMD.  The obsid of the most recent maneuver is
+    also captured, as it looks like for science processing the problem can be that
+    the gradient data for the science interval extends into the time of an anomaly,
+    even though AOACASEQ and AOPCADMD show that the spacecraft is in anomaly state,
+    and COBSRQID is often 0.
+
+    Parameters
+    ----------
+    start : str
+        Start time for search
+
+    Returns
+    -------
+    bad_events : astropy.table.Table
+        Table of bad events
+    """
+    msids = ["OOBAGRD3", "OOBAGRD6"]
+    quantization = 0.0032
+    sampling = 32.80000185966492
+    quant_mult = 5
+
+    bad_events = []
+    for msid in msids:
+        dat = fetch.Msid(msid, start, filter_bad=True)
+        dat.interpolate(dt=sampling)
+        diff = np.diff(dat.vals)
+        nok = np.abs(diff) > quantization * quant_mult
+
+        for hit in np.where(nok)[0]:
+            event = {
+                "time": dat.times[hit],
+                "date": CxoTime(dat.times[hit]).date,
+                "msid": msid,
+            }
+            local_manvrs = events.manvrs.filter(
+                start=dat.times[hit] - (7 * 86400), stop=dat.times[hit] + (1 * 86400)
+            )
+            extra_msids = ["COBSRQID", "AOPCADMD", "AOACASEQ"]
+            extra_defaults = {"COBSRQID": 0, "AOPCADMD": "N/A", "AOACASEQ": "N/A"}
+            for extra_msid in extra_msids:
+                mdat = fetch.Msid(extra_msid, dat.times[hit] - 200, dat.times[hit] + 50)
+                # get last sample before hit
+                hit_idx = np.searchsorted(mdat.times, dat.times[hit])
+                if hit_idx == 0:
+                    event.update({extra_msid: extra_defaults[extra_msid]})
+                else:
+                    event.update({extra_msid: mdat.vals[hit_idx - 1]})
+            last_manvr = None
```

Comment:
Not sure about the cleanest method to just find the last expected obsid for the commanded maneuver/dwell.  This is a little convoluted but seems to work robustly when there is bad or missing data (including kadi events based on telemetry).